Possible issues for `PipelineExecGroupBuilder`

In testing, logs were found indicating that there may have been some exceptions occurring with `PipelineExecGroupBuilder`. 
Although the system can still function properly even if `PipelineExecGroupBuilder` throws an error here, it is necessary to add checks.

```
[2023/05/29 19:54:08.328 +08:00] [ERROR] [BaseDaemon.cpp:569] ["
       0xacd5e1d\tfaultSignalHandler(int, siginfo_t*, void*) [tiflash+181231133]
                \tlibs/libdaemon/src/BaseDaemon.cpp:220
  0x7f828f3acd90\t<unknown symbol> [libc.so.6+347536]
  0x7f828f3f954c\t__pthread_kill_implementation [libc.so.6+660812]
  0x7f828f3acce6\tgsignal [libc.so.6+347366]
  0x7f828f3807f3\t__GI_abort [libc.so.6+165875]
  0x7f829a19f66f\tstd::__1::__libcpp_abort_debug_function(std::__1::__libcpp_debug_info const&) [libc++.so.1+464495]
       0xaa36122\tstd::__1::vector<std::__1::vector<DB::PipelineExecBuilder, std::__1::allocator<DB::PipelineExecBuilder> >, std::__1::allocator<std::__1::vector<DB::PipelineExecBuilder, std::__1::allocator<DB::PipelineExecBuilder> > > >::back() [tiflash+178479394]
                \t/usr/local/bin/../include/c++/v1/vector:708
       0xaa35f55\tDB::PipelineExecGroupBuilder::getCurGroup() [tiflash+178478933]
                \tdbms/src/Flash/Pipeline/Exec/PipelineExecBuilder.h:46
       0xbfa8bfd\tDB::PipelineExecGroupBuilder::addConcurrency(std::__1::unique_ptr<DB::SourceOp, std::__1::default_delete<DB::SourceOp> >&&) [tiflash+200969213]
                \tdbms/src/Flash/Pipeline/Exec/PipelineExecBuilder.cpp:63
       0xaa00403\tDB::DM::DeltaMergeStore::read(DB::PipelineExecutorStatus&, DB::PipelineExecGroupBuilder&, DB::Context const&, DB::Settings const&, std::__1::vector<DB::DM::ColumnDefine, std::__1::allocator<DB::DM::ColumnDefine> > const&, std::__1::vector<DB::DM::RowKeyRange, std::__1::allocator<DB::DM::RowKeyRange> > const&, unsigned long, unsigned long, std::__1::shared_ptr<DB::DM::PushDownFilter> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, bool, bool, unsigned long, std::__1::unordered_set<unsigned long, std::__1::hash<unsigned long>, std::__1::equal_to<unsigned long>, std::__1::allocator<unsigned long> > const&, unsigned long, std::__1::shared_ptr<DB::DM::ScanContext>) [tiflash+178258947]
                \tdbms/src/Storages/DeltaMerge/DeltaMergeStore.cpp:1193
       0xb93707e\tDB::StorageDeltaMerge::read(DB::PipelineExecutorStatus&, DB::PipelineExecGroupBuilder&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, DB::SelectQueryInfo const&, DB::Context const&, unsigned long, unsigned int) [tiflash+194211966]
                \tdbms/src/Storages/StorageDeltaMerge.cpp:972
       0xbdea0c9\tDB::DAGStorageInterpreter::buildLocalExecForPhysicalTable(DB::PipelineExecutorStatus&, DB::PipelineExecGroupBuilder&, long const&, DB::SelectQueryInfo const&, unsigned long) [tiflash+199139529]
                \tdbms/src/Flash/Coprocessor/DAGStorageInterpreter.cpp:953
       0xbddfedf\tDB::DAGStorageInterpreter::buildLocalExec(DB::PipelineExecutorStatus&, DB::PipelineExecGroupBuilder&, unsigned long) [tiflash+199098079]
                \tdbms/src/Flash/Coprocessor/DAGStorageInterpreter.cpp:1091
       0xbddf58b\tDB::DAGStorageInterpreter::executeImpl(DB::PipelineExecutorStatus&, DB::PipelineExecGroupBuilder&) [tiflash+199095691]
                \tdbms/src/Flash/Coprocessor/DAGStorageInterpreter.cpp:296
       0xbddf491\tDB::DAGStorageInterpreter::execute(DB::PipelineExecutorStatus&, DB::PipelineExecGroupBuilder&) [tiflash+199095441]
                \tdbms/src/Flash/Coprocessor/DAGStorageInterpreter.cpp:283
       0xc04ebf2\tDB::PhysicalTableScan::buildPipeline(DB::PipelineBuilder&, DB::Context&, DB::PipelineExecutorStatus&) [tiflash+201649138]
                \tdbms/src/Flash/Planner/Plans/PhysicalTableScan.cpp:90
       0xbfdf02a\tDB::PhysicalPlanNode::buildPipeline(DB::PipelineBuilder&, DB::Context&, DB::PipelineExecutorStatus&) [tiflash+201191466]
                \tdbms/src/Flash/Planner/PhysicalPlanNode.cpp:107
       0xc02cd47\tDB::PhysicalJoin::buildPipeline(DB::PipelineBuilder&, DB::Context&, DB::PipelineExecutorStatus&) [tiflash+201510215]
                \tdbms/src/Flash/Planner/Plans/PhysicalJoin.cpp:264
       0xbfdf02a\tDB::PhysicalPlanNode::buildPipeline(DB::PipelineBuilder&, DB::Context&, DB::PipelineExecutorStatus&) [tiflash+201191466]
                \tdbms/src/Flash/Planner/PhysicalPlanNode.cpp:107
       0xbfdf02a\tDB::PhysicalPlanNode::buildPipeline(DB::PipelineBuilder&, DB::Context&, DB::PipelineExecutorStatus&) [tiflash+201191466]
                \tdbms/src/Flash/Planner/PhysicalPlanNode.cpp:107
       0xbfd88b3\tDB::PhysicalPlan::toPipeline(DB::PipelineExecutorStatus&, DB::Context&) [tiflash+201164979]
                \tdbms/src/Flash/Planner/PhysicalPlan.cpp:285
       0xbf8543f\tDB::PipelineExecutor::PipelineExecutor(std::__1::shared_ptr<MemoryTracker> const&, DB::Context&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) [tiflash+200823871]
                \tdbms/src/Flash/Executor/PipelineExecutor.cpp:34
       0xbd66541\tstd::__1::__unique_if<DB::PipelineExecutor>::__unique_single std::__1::make_unique<DB::PipelineExecutor, std::__1::shared_ptr<MemoryTracker>&, DB::Context&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&>(std::__1::shared_ptr<MemoryTracker>&, DB::Context&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) [tiflash+198600001]
                \t/usr/local/bin/../include/c++/v1/__memory/unique_ptr.h:728
       0xbd65d78\tDB::(anonymous namespace)::executeAsPipeline(DB::Context&, bool) [tiflash+198598008]
                \tdbms/src/Flash/executeQuery.cpp:143
       0xbd65910\tDB::queryExecute(DB::Context&, bool) [tiflash+198596880]
                \tdbms/src/Flash/executeQuery.cpp:179
       0xbefbafc\tDB::MPPTask::preprocess() [tiflash+200260348]
                \tdbms/src/Flash/Mpp/MPPTask.cpp:361
       0xbefc3e3\tDB::MPPTask::runImpl() [tiflash+200262627]
                \tdbms/src/Flash/Mpp/MPPTask.cpp:403
       0xbf015fd\tDB::MPPTask::run()::$_17::operator()() const [tiflash+200283645]
                \tdbms/src/Flash/Mpp/MPPTask.cpp:154
       0xbf015c5\tdecltype(static_cast<DB::MPPTask::run()::$_17&>(fp)()) std::__1::__invoke<DB::MPPTask::run()::$_17&>(DB::MPPTask::run()::$_17&) [tiflash+200283589]
                \t/usr/local/bin/../include/c++/v1/type_traits:3918
       0xbf0157d\tvoid std::__1::__invoke_void_return_wrapper<void, true>::__call<DB::MPPTask::run()::$_17&>(DB::MPPTask::run()::$_17&) [tiflash+200283517]
                \t/usr/local/bin/../include/c++/v1/__functional/invoke.h:61
       0xbf0154d\tstd::__1::__function::__alloc_func<DB::MPPTask::run()::$_17, std::__1::allocator<DB::MPPTask::run()::$_17>, void ()>::operator()() [tiflash+200283469]
                \t/usr/local/bin/../include/c++/v1/__functional/function.h:171
       0xbf00689\tstd::__1::__function::__func<DB::MPPTask::run()::$_17, std::__1::allocator<DB::MPPTask::run()::$_17>, void ()>::operator()() [tiflash+200279689]
                \t/usr/local/bin/../include/c++/v1/__functional/function.h:345
       0x43758f6\tstd::__1::__function::__value_func<void ()>::operator()() const [tiflash+70736118]
                \t/usr/local/bin/../include/c++/v1/__functional/function.h:498
       0x436e405\tstd::__1::function<void ()>::operator()() const [tiflash+70706181]
                \t/usr/local/bin/../include/c++/v1/__functional/function.h:1175
       0x437ff55\tdecltype(static_cast<std::__1::function<void ()>>(fp)()) std::__1::__invoke_constexpr<std::__1::function<void ()> >(std::__1::function<void ()>&&) [tiflash+70778709]
                \t/usr/local/bin/../include/c++/v1/type_traits:3924
       0x437ff31\tdecltype(auto) std::__1::__apply_tuple_impl<std::__1::function<void ()>, std::__1::tuple<> >(std::__1::function<void ()>&&, std::__1::tuple<>&&, std::__1::__tuple_indices<>) [tiflash+70778673]
                \t/usr/local/bin/../include/c++/v1/tuple:1536
       0x437ff01\tdecltype(auto) std::__1::apply<std::__1::function<void ()>, std::__1::tuple<> >(std::__1::function<void ()>&&, std::__1::tuple<>&&) [tiflash+70778625]
                \t/usr/local/bin/../include/c++/v1/tuple:1545
       0x437fe9c\tauto DB::wrapInvocable<std::__1::function<void ()> >(bool, std::__1::function<void ()>&&)::'lambda'()::operator()() [tiflash+70778524]
                \tdbms/src/Common/wrapInvocable.h:36
       0x4380b69\tDB::ExecutableTask<auto DB::wrapInvocable<std::__1::function<void ()> >(bool, std::__1::function<void ()>&&)::'lambda'()>::execute() [tiflash+70781801]
                \tdbms/src/Common/ExecutableTask.h:52
       0x427b29d\tDB::DynamicThreadPool::executeTask(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&) [tiflash+69710493]
                \tdbms/src/Common/DynamicThreadPool.cpp:101
       0x427aabf\tDB::DynamicThreadPool::fixedWork(unsigned long) [tiflash+69708479]
                \tdbms/src/Common/DynamicThreadPool.cpp:115
       0x42835cc\tdecltype(*(static_cast<DB::DynamicThreadPool*>(fp0)).*fp(static_cast<unsigned long&>(fp1))) std::__1::__invoke<void (DB::DynamicThreadPool::* const&)(unsigned long), DB::DynamicThreadPool*, unsigned long&, void>(void (DB::DynamicThreadPool::* const&)(unsigned long), DB::DynamicThreadPool*&&, unsigned long&) [tiflash+69744076]
                \t/usr/local/bin/../include/c++/v1/type_traits:3859
       0x42834d6\tstd::__1::invoke_result<void (DB::DynamicThreadPool::* const&)(unsigned long), DB::DynamicThreadPool*, unsigned long&>::type std::__1::invoke<void (DB::DynamicThreadPool::* const&)(unsigned long), DB::DynamicThreadPool*, unsigned long&>(void (DB::DynamicThreadPool::* const&)(unsigned long), DB::DynamicThreadPool*&&, unsigned long&) [tiflash+69743830]
                \t/usr/local/bin/../include/c++/v1/__functional/invoke.h:93
       0x42833c5\tauto std::__1::thread DB::ThreadFactory::newThread<void (DB::DynamicThreadPool::*)(unsigned long), DB::DynamicThreadPool*, unsigned long&>(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void (DB::DynamicThreadPool::*&&)(unsigned long), DB::DynamicThreadPool*&&, unsigned long&)::'lambda'(auto&&...)::operator()<DB::DynamicThreadPool*, unsigned long>(auto&&...) const [tiflash+69743557]
                \tdbms/src/Common/ThreadFactory.h:46
       0x42831f5\tdecltype(static_cast<void (DB::DynamicThreadPool::*>(fp)(static_cast<DB::DynamicThreadPool*>(fp0), static_cast<unsigned long&>(fp0))) std::__1::__invoke<std::__1::thread DB::ThreadFactory::newThread<void (DB::DynamicThreadPool::*)(unsigned long), DB::DynamicThreadPool*, unsigned long&>(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void (DB::DynamicThreadPool::*&&)(unsigned long), DB::DynamicThreadPool*&&, unsigned long&)::'lambda'(auto&&...), DB::DynamicThreadPool*, unsigned long>(void (DB::DynamicThreadPool::*&&)(unsigned long), DB::DynamicThreadPool*&&, unsigned long&) [tiflash+69743093]
                \t/usr/local/bin/../include/c++/v1/type_traits:3918
       0x4283146\tvoid std::__1::__thread_execute<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, std::__1::thread DB::ThreadFactory::newThread<void (DB::DynamicThreadPool::*)(unsigned long), DB::DynamicThreadPool*, unsigned long&>(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void (DB::DynamicThreadPool::*&&)(unsigned long), DB::DynamicThreadPool*&&, unsigned long&)::'lambda'(auto&&...), DB::DynamicThreadPool*, unsigned long, 2ul, 3ul>(std::__1::tuple<void (DB::DynamicThreadPool::*)(unsigned long), std::__1::thread DB::ThreadFactory::newThread<void (DB::DynamicThreadPool::*)(unsigned long), DB::DynamicThreadPool*, unsigned long&>(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void (DB::DynamicThreadPool::*&&)(unsigned long), DB::DynamicThreadPool*&&, unsigned long&)::'lambda'(auto&&...), DB::DynamicThreadPool*, unsigned long>&, std::__1::__tuple_indices<2ul, 3ul>) [tiflash+69742918]
                \t/usr/local/bin/../include/c++/v1/thread:280
       0x4282a4b\tvoid* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, std::__1::thread DB::ThreadFactory::newThread<void (DB::DynamicThreadPool::*)(unsigned long), DB::DynamicThreadPool*, unsigned long&>(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void (DB::DynamicThreadPool::*&&)(unsigned long), DB::DynamicThreadPool*&&, unsigned long&)::'lambda'(auto&&...), DB::DynamicThreadPool*, unsigned long> >(void*) [tiflash+69741131]
                \t/usr/local/bin/../include/c++/v1/thread:291
  0x7f828f3f7802\tstart_thread [libc.so.6+653314]"] [source=BaseDaemon] [thread_id=1011]
```

--- a/dbms/src/Flash/Coprocessor/DAGStorageInterpreter.cpp
+++ b/dbms/src/Flash/Coprocessor/DAGStorageInterpreter.cpp
@@ -513,8 +513,8 @@ void DAGStorageInterpreter::executeCastAfterTableScan(
         // local sources
         while (i < remote_read_start_index)
         {
-            auto & group = group_builder.getCurGroup()[i++];
-            group.appendTransformOp(std::make_unique<ExpressionTransformOp>(exec_status, log->identifier(), extra_cast));
+            auto & builder = group_builder.getCurBuilder(i++);
+            builder.appendTransformOp(std::make_unique<ExpressionTransformOp>(exec_status, log->identifier(), extra_cast));
         }
     }
 }

--- a/dbms/src/Flash/Coprocessor/InterpreterUtils.cpp
+++ b/dbms/src/Flash/Coprocessor/InterpreterUtils.cpp
@@ -406,10 +406,10 @@ void executePushedDownFilter(
     // for remote read, filter had been pushed down, don't need to execute again.
     for (size_t i = 0; i < remote_read_sources_start_index; ++i)
     {
-        auto & group = group_builder.getCurGroup()[i];
-        group.appendTransformOp(std::make_unique<FilterTransformOp>(exec_status, log->identifier(), input_header, before_where, filter_column_name));
+        auto & builder = group_builder.getCurBuilder(i);
+        builder.appendTransformOp(std::make_unique<FilterTransformOp>(exec_status, log->identifier(), input_header, before_where, filter_column_name));
         // after filter, do project action to keep the schema of local transforms and remote transforms the same.
-        group.appendTransformOp(std::make_unique<ExpressionTransformOp>(exec_status, log->identifier(), project_after_where));
+        builder.appendTransformOp(std::make_unique<ExpressionTransformOp>(exec_status, log->identifier(), project_after_where));
     }
 }
 
@@ -465,8 +465,8 @@ void executeGeneratedColumnPlaceholder(
 
     for (size_t i = 0; i < remote_read_sources_start_index; ++i)
     {
-        auto & group = group_builder.getCurGroup()[i];
-        group.appendTransformOp(std::make_unique<GeneratedColumnPlaceHolderTransformOp>(exec_status, log->identifier(), group_builder.getCurrentHeader(), generated_column_infos));
+        auto & builder = group_builder.getCurBuilder(i);
+        builder.appendTransformOp(std::make_unique<GeneratedColumnPlaceHolderTransformOp>(exec_status, log->identifier(), group_builder.getCurrentHeader(), generated_column_infos));
     }
 }
 

--- a/dbms/src/Flash/Pipeline/Exec/PipelineExecBuilder.cpp
+++ b/dbms/src/Flash/Pipeline/Exec/PipelineExecBuilder.cpp
@@ -67,6 +67,8 @@ void PipelineExecGroupBuilder::addConcurrency(SourceOpPtr && source)
 void PipelineExecGroupBuilder::reset()
 {
     groups.clear();
+    // Re-add an empty group to ensure that the build after reset is available.
+    groups.emplace_back();
 }
 
 void PipelineExecGroupBuilder::merge(PipelineExecGroupBuilder && other)

--- a/dbms/src/Flash/Pipeline/Exec/PipelineExecBuilder.cpp
+++ b/dbms/src/Flash/Pipeline/Exec/PipelineExecBuilder.cpp
@@ -60,14 +60,15 @@ Block PipelineExecBuilder::getCurrentHeader() const
 
 void PipelineExecGroupBuilder::addConcurrency(SourceOpPtr && source)
 {
-    getCurGroup().emplace_back();
-    getCurGroup().back().setSourceOp(std::move(source));
+    auto & cur_group = getCurGroup();
+    cur_group.emplace_back();
+    cur_group.back().setSourceOp(std::move(source));
 }
 
 void PipelineExecGroupBuilder::reset()
 {
     groups.clear();
-    // Re-add an empty group to ensure that the build after reset is available.
+    // Re-add an empty group to ensure that the group builder after reset is available.
     groups.emplace_back();
 }
 
@@ -94,7 +95,8 @@ PipelineExecGroup PipelineExecGroupBuilder::build()
 
 Block PipelineExecGroupBuilder::getCurrentHeader()
 {
-    RUNTIME_CHECK(!getCurGroup().empty());
-    return getCurGroup().back().getCurrentHeader();
+    auto & cur_group = getCurGroup();
+    RUNTIME_CHECK(!cur_group.empty());
+    return cur_group.back().getCurrentHeader();
 }
 } // namespace DB

--- a/dbms/src/Flash/Pipeline/Exec/PipelineExecBuilder.h
+++ b/dbms/src/Flash/Pipeline/Exec/PipelineExecBuilder.h
@@ -43,12 +43,6 @@ public:
 
     using BuilderGroup = std::vector<PipelineExecBuilder>;
 
-    BuilderGroup & getCurGroup()
-    {
-        RUNTIME_CHECK(!groups.empty());
-        return groups.back();
-    }
-
     PipelineExecBuilder & getCurBuilder(size_t index)
     {
         RUNTIME_CHECK(!groups.empty());
@@ -81,6 +75,13 @@ public:
     PipelineExecGroup build();
 
     Block getCurrentHeader();
+
+private:
+    BuilderGroup & getCurGroup()
+    {
+        RUNTIME_CHECK(!groups.empty());
+        return groups.back();
+    }
 
 private:
     // groups generates a set of pipeline_execs running in parallel.

--- a/dbms/src/Flash/Pipeline/Exec/PipelineExecBuilder.h
+++ b/dbms/src/Flash/Pipeline/Exec/PipelineExecBuilder.h
@@ -43,7 +43,18 @@ public:
 
     using BuilderGroup = std::vector<PipelineExecBuilder>;
 
-    BuilderGroup & getCurGroup() { return groups.back(); }
+    BuilderGroup & getCurGroup()
+    {
+        RUNTIME_CHECK(!groups.empty());
+        return groups.back();
+    }
+
+    PipelineExecBuilder & getCurBuilder(size_t index)
+    {
+        RUNTIME_CHECK(!groups.empty());
+        RUNTIME_CHECK(groups.back().size() > index);
+        return groups.back()[index];
+    }
 
     void addGroup() { groups.emplace_back(); }
 

